assets/js/background.js

```diff
@@ -178,6 +185,103 @@ function handleOnUpdated(tabId, changeInfo, tab) {
 }
 
 /**
+ * Message listenener for context requests
+ */
+var findContext = function findContext(request, sender, sendResponse) {
+  chrome.tabs.query({
+    active: true,
+    currentWindow: true
+  }, function(tabs) {
+    var tab = tabs[0];
+    var site = getSiteFromTab(config.supportedSites, tab);
+    // Unsupported website, skip.
+    if(!site) {
+      console.warn('Unsupported');
+      return;
+    }
+    // Retrieve context from tab
+    detectContextWithRetry(tab, site, 2, 1000, function(err, context) {
+      if(err) {
+        console.warn(err);
+        return;
+      }
+      sendResponse({
+        site: site,
+        context: blacklist.filterQuery(context)
+      });
+      return;
+    });
+  });
+  return true;
+};
+
+
+/**
+ * Message listenener for search requests
+ */
+var getResults = function getResults(request, sender, sendResponse) {
+  var context = request.context;
+  if(!context.length) {
+    return;
+  }
+
+  var query = generateQuery(context);
+  var payload = {
+    query: query,
+    check: request.check
+  };
+
+  async.waterfall([
+    function loadSettings(cb) {
+      if(config.loaded) {
+        return cb(null);
+      }
+      config.loadUserSettings(cb);
+    },
+    function search(cb) {
+      getDocuments(query, cb);
+    },
+    function respond(documents, count, cb) {
+      payload.count = count;
+      payload.documents = documents;
+      sendResponse(payload);
+      cb(null);
+    }
+  ], function(err) {
+    if(err) {
+      console.warn(err);
+    }
+  });
+
+  return true;
```

Comment:
From the Chrome extension doc:
>**sendResponse**
>Function to call (at most once) when you have a response. The argument should be any JSON-ifiable object. If you have more than one onMessage listener in the same document, then only one may send a response. This function becomes invalid when the event listener returns, **unless you return true from the event listener to indicate you wish to send a response asynchronously** (this will keep the message channel open to the other end until sendResponse is called).

https://developer.chrome.com/extensions/runtime


This might deserve a comment however